[SPARK-13579][build][test-maven] Stop building the main Spark assembly.

vanzin opened the pull request:
This change modifies the "assembly/" module to just copy needed
dependencies to its build directory, and modifies the packaging
script to pick those up (and remove duplicate jars packages in the
examples module).

I also made some minor adjustments to dependencies to remove some
test jars from the final packaging, and remove jars that conflict with each
other when packaged separately (e.g. servlet api).

Also note that this change restores guava in applications' classpaths, even
though it's still shaded inside Spark. This is now needed for the Hadoop
libraries that are packaged with Spark, which now are not processed by
the shade plugin.

Commit:
[SPARK-13579][build][wip] Stop building the main Spark assembly.

This change modifies the "assembly/" module to just copy needed
dependencies to its build directory, and modifies the packaging
script to pick those up (and remove duplicate jars packages in the
examples module).

I also made some minor adjustments to dependencies to remove some
test jars from the final packaging.

vanzin commented:
This is probably not ready yet, although it works for me locally. I want to run the patch through jenkins (sbt and maven) to see how well it works out.

SparkQA commented:
**[Test build #53471 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53471/consoleFull)** for PR 11796 at commit [`f679e14`](https://github.com/apache/spark/commit/f679e142f884f0318ddec12e9b5c3b62433e303a).

SparkQA commented:
**[Test build #53471 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53471/consoleFull)** for PR 11796 at commit [`f679e14`](https://github.com/apache/spark/commit/f679e142f884f0318ddec12e9b5c3b62433e303a).
 * This patch **fails build dependency tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53471/
Test FAILed.

Commit:
Update dependencies.

SparkQA commented:
**[Test build #53472 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53472/consoleFull)** for PR 11796 at commit [`54336b6`](https://github.com/apache/spark/commit/54336b6305a0cc5fb2b80247f5ab76e1b6f08407).

Commit:
Make guava use same scope as hadoop.

SparkQA commented:
**[Test build #53473 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53473/consoleFull)** for PR 11796 at commit [`5d1e259`](https://github.com/apache/spark/commit/5d1e259a9659987dd14da54adf1f168220bdafa2).

hvanhovell commented on sql/hive/pom.xml:
```diff
@@ -238,36 +238,13 @@
                <sources>
                  <source>v${hive.version.short}/src/main/scala</source>
-                 <source>${project.build.directory/generated-sources/antlr</source>
+                 <source>${project.build.directory}/generated-sources/antlr</source>
```
Nice catch :)

vanzin replied:
It didn't seem to be affecting the build, though. I'm actually wondering if this line is needed - the parser seems to be part of the catalyst module, not this one.

hvanhovell replied:
Yeah you are right about that. So this could actually be removed. Lemme know, I can also submit a PR for this.

vanzin replied:
I'll remove it, but since this patch may take some more iterations before it's ready, feel free to do it as part of another patch too.

SparkQA commented:
**[Test build #53473 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53473/consoleFull)** for PR 11796 at commit [`5d1e259`](https://github.com/apache/spark/commit/5d1e259a9659987dd14da54adf1f168220bdafa2).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53473/
Test FAILed.

SparkQA commented:
**[Test build #53472 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53472/consoleFull)** for PR 11796 at commit [`54336b6`](https://github.com/apache/spark/commit/54336b6305a0cc5fb2b80247f5ab76e1b6f08407).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53472/
Test FAILed.

Commit:
Fix tests (contains fix borrowed from #11701).

SparkQA commented:
**[Test build #53549 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53549/consoleFull)** for PR 11796 at commit [`5ec6722`](https://github.com/apache/spark/commit/5ec6722eb33d72c24e7a2b25002f6161f9cb4aa2).

SparkQA commented:
**[Test build #53549 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53549/consoleFull)** for PR 11796 at commit [`5ec6722`](https://github.com/apache/spark/commit/5ec6722eb33d72c24e7a2b25002f6161f9cb4aa2).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53549/
Test FAILed.

Commit:
Fix a test that might suppress further logging when it fails.

SparkQA commented:
**[Test build #53576 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53576/consoleFull)** for PR 11796 at commit [`a83afeb`](https://github.com/apache/spark/commit/a83afebca19ac4bb9a2c51f3da823924daef1554).

SparkQA commented:
**[Test build #53576 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53576/consoleFull)** for PR 11796 at commit [`a83afeb`](https://github.com/apache/spark/commit/a83afebca19ac4bb9a2c51f3da823924daef1554).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53576/
Test FAILed.

vanzin commented:
Hmm, log files are gone and tests pass locally... sigh. retest this please

SparkQA commented:
**[Test build #53683 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53683/consoleFull)** for PR 11796 at commit [`a83afeb`](https://github.com/apache/spark/commit/a83afebca19ac4bb9a2c51f3da823924daef1554).

SparkQA commented:
**[Test build #53683 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53683/consoleFull)** for PR 11796 at commit [`a83afeb`](https://github.com/apache/spark/commit/a83afebca19ac4bb9a2c51f3da823924daef1554).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53683/
Test FAILed.

Commit:
Fix thrift server tests.

These tests disable "SPARK_TESTING" for whatever reason; so add a new
env variable that's specific to them, and handle that in the launch
scripts.

SparkQA commented:
**[Test build #53735 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53735/consoleFull)** for PR 11796 at commit [`5d517f0`](https://github.com/apache/spark/commit/5d517f066ab0adc4c189c09ce78b5e8232dd3792).

SparkQA commented:
**[Test build #53735 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53735/consoleFull)** for PR 11796 at commit [`5d517f0`](https://github.com/apache/spark/commit/5d517f066ab0adc4c189c09ce78b5e8232dd3792).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53735/
Test FAILed.

vanzin commented:
These tests fail sporadically; I think it's some weird sbt dependency resolution issue that's causing different hadoop versions to get mixed up. e.g.

```
16/03/21 19:43:11.830 redirect stderr for command ./bin/spark-submit INFO Utils: Exception in thread "main" java.lang.NoSuchMethodError: org.apache.hadoop.conf.Configuration.addDeprecations([Lorg/apache/hadoop/conf/Configuration$DeprecationDelta;)V
16/03/21 19:43:11.830 redirect stderr for command ./bin/spark-submit INFO Utils: 	at org.apache.hadoop.mapreduce.util.ConfigUtil.addDeprecatedKeys(ConfigUtil.java:54)
16/03/21 19:43:11.830 redirect stderr for command ./bin/spark-submit INFO Utils: 	at org.apache.hadoop.mapreduce.util.ConfigUtil.loadResources(ConfigUtil.java:42)
16/03/21 19:43:11.830 redirect stderr for command ./bin/spark-submit INFO Utils: 	at org.apache.hadoop.mapred.JobConf.<clinit>(JobConf.java:118)
```

That's not the whole stack trace, but there's only hadoop stuff in there, so it's not Spark calling some method that was removed.

vanzin commented:
The offending method (`void addDeprecations(DeprecationDelta[])`) seems to have been added in Hadoop 2.3 and seems to exist in all versions since. Somehow we're getting hadoop 2.2 jars in the classpath?

SparkQA commented:
**[Test build #53800 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53800/consoleFull)** for PR 11796 at commit [`83fa3a4`](https://github.com/apache/spark/commit/83fa3a4c1445532bd9115bd841d25f51185ec2ca).

SparkQA commented:
**[Test build #53800 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53800/consoleFull)** for PR 11796 at commit [`83fa3a4`](https://github.com/apache/spark/commit/83fa3a4c1445532bd9115bd841d25f51185ec2ca).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53800/
Test FAILed.

vanzin commented:
Weird jline error, trying to see if I can figure it out... but the tests I want to fail passed. retest this please.

SparkQA commented:
**[Test build #53818 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53818/consoleFull)** for PR 11796 at commit [`83fa3a4`](https://github.com/apache/spark/commit/83fa3a4c1445532bd9115bd841d25f51185ec2ca).

vanzin commented:
jline error looks weirdly similar. Old version of jline (0.9) has class `jline.Terminal`, while newer versions have an interface. 

```
2016-03-22 13:00:57.304 - stderr> java.lang.IncompatibleClassChangeError: Found class jline.Terminal, but interface was expected
2016-03-22 13:00:57.306 - stderr> 	at jline.TerminalFactory.create(TerminalFactory.java:101)
2016-03-22 13:00:57.306 - stderr> 	at jline.TerminalFactory.get(TerminalFactory.java:158)
2016-03-22 13:00:57.306 - stderr> 	at jline.console.ConsoleReader.<init>(ConsoleReader.java:229)
2016-03-22 13:00:57.306 - stderr> 	at jline.console.ConsoleReader.<init>(ConsoleReader.java:221)
```

SparkQA commented:
**[Test build #53818 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53818/consoleFull)** for PR 11796 at commit [`83fa3a4`](https://github.com/apache/spark/commit/83fa3a4c1445532bd9115bd841d25f51185ec2ca).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53818/
Test FAILed.

vanzin commented:
I'm a little at a loss at what's wrong. Copy & pasting the commands from the logs and running them *on the same jenkins machine* works (I don't get that exception). Running `javap` for the Configuration class with the command's classpath shows the new class, with the method that the logs complain about...

SparkQA commented:
**[Test build #53871 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53871/consoleFull)** for PR 11796 at commit [`0339263`](https://github.com/apache/spark/commit/03392634cfe27064100479ea92fa0d4d465f3aab).

SparkQA commented:
**[Test build #53874 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53874/consoleFull)** for PR 11796 at commit [`b5b8dd1`](https://github.com/apache/spark/commit/b5b8dd1833106f8f267a21f80108a00bde8cfad1).

SparkQA commented:
**[Test build #53874 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53874/consoleFull)** for PR 11796 at commit [`b5b8dd1`](https://github.com/apache/spark/commit/b5b8dd1833106f8f267a21f80108a00bde8cfad1).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53874/
Test FAILed.

SparkQA commented:
**[Test build #53875 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53875/consoleFull)** for PR 11796 at commit [`6c4ed0a`](https://github.com/apache/spark/commit/6c4ed0a2ea0b3ae968b4d5a8a81555ea3e4cbcf4).

SparkQA commented:
**[Test build #53871 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53871/consoleFull)** for PR 11796 at commit [`0339263`](https://github.com/apache/spark/commit/03392634cfe27064100479ea92fa0d4d465f3aab).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53871/
Test FAILed.

SparkQA commented:
**[Test build #53875 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53875/consoleFull)** for PR 11796 at commit [`6c4ed0a`](https://github.com/apache/spark/commit/6c4ed0a2ea0b3ae968b4d5a8a81555ea3e4cbcf4).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53875/
Test FAILed.

Commit:
Set needed classpath for python tests.

SparkQA commented:
**[Test build #53948 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53948/consoleFull)** for PR 11796 at commit [`e37886b`](https://github.com/apache/spark/commit/e37886badf3c4fd5da239aa5cf1d232ddb14043b).

SparkQA commented:
**[Test build #53948 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53948/consoleFull)** for PR 11796 at commit [`e37886b`](https://github.com/apache/spark/commit/e37886badf3c4fd5da239aa5cf1d232ddb14043b).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53948/
Test FAILed.

SparkQA commented:
**[Test build #53949 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53949/consoleFull)** for PR 11796 at commit [`55aec21`](https://github.com/apache/spark/commit/55aec212c35801621cf7fe57e920b58fd5d29eb2).

SparkQA commented:
**[Test build #53949 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53949/consoleFull)** for PR 11796 at commit [`55aec21`](https://github.com/apache/spark/commit/55aec212c35801621cf7fe57e920b58fd5d29eb2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53949/
Test PASSed.

vanzin commented:
yay, but I'm still worried about the previous failures. will test on yarn now.

vanzin commented:
retest this please

SparkQA commented:
**[Test build #53963 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53963/consoleFull)** for PR 11796 at commit [`55aec21`](https://github.com/apache/spark/commit/55aec212c35801621cf7fe57e920b58fd5d29eb2).

Commit:
Remove unneeded line.

vanzin commented:
the maven build seems stuck... :-/ I'll revert to sbt and add some patches since I want to try to hit the previous failures.

SparkQA commented:
**[Test build #53981 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53981/consoleFull)** for PR 11796 at commit [`5c660aa`](https://github.com/apache/spark/commit/5c660aac3b6d44622ab39656367a6f45b0b46402).

SparkQA commented:
**[Test build #53963 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53963/consoleFull)** for PR 11796 at commit [`55aec21`](https://github.com/apache/spark/commit/55aec212c35801621cf7fe57e920b58fd5d29eb2).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53963/
Test FAILed.

SparkQA commented:
**[Test build #53981 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53981/consoleFull)** for PR 11796 at commit [`5c660aa`](https://github.com/apache/spark/commit/5c660aac3b6d44622ab39656367a6f45b0b46402).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53981/
Test PASSed.

vanzin commented:
retest this please

SparkQA commented:
**[Test build #53992 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53992/consoleFull)** for PR 11796 at commit [`5c660aa`](https://github.com/apache/spark/commit/5c660aac3b6d44622ab39656367a6f45b0b46402).

Commit:
Fix regression: allow tests to run without the need to run "package" first.

SparkQA commented:
**[Test build #53992 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53992/consoleFull)** for PR 11796 at commit [`5c660aa`](https://github.com/apache/spark/commit/5c660aac3b6d44622ab39656367a6f45b0b46402).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53992/
Test FAILed.

Commit:
Fix for SPARK-14134 (#11941).

vanzin commented:
Aha. Somehow an old jline is slipping into the final packaging; from the latest CliSuite failures, this was in the JVM's classpath:

    /home/jenkins/workspace/SparkPullRequestBuilder/assembly/target/scala-2.11/jars/jline-0.9.94.jar

And sure enough there's an old Hadoop version there too:

    /home/jenkins/workspace/SparkPullRequestBuilder/assembly/target/scala-2.11/jars/hadoop-mapreduce-client-shuffle-2.2.0.jar

Now need to figure out why that's happening.

srowen commented:
PS that sounds like the cause of https://issues.apache.org/jira/browse/SPARK-13710, at least as the JIRA describes it, though the old jline is coming from curator-recipes. I had suggested excluding it as the fix there.

vanzin commented:
> I had suggested excluding it as the fix there.

The thing is that I don't see it in any of the dependency graphs (maven or sbt); also, that doesn't explain why two different sets of hadoop jars are showing up (the classpath had both 2.2.0 and 2.3.0 versions of the same hadoop artifacts!).

SparkQA commented:
**[Test build #54078 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54078/consoleFull)** for PR 11796 at commit [`4fdfc4e`](https://github.com/apache/spark/commit/4fdfc4e0cf9994d0d69e2b74daefc36d22a090ce).

SparkQA commented:
**[Test build #54078 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54078/consoleFull)** for PR 11796 at commit [`4fdfc4e`](https://github.com/apache/spark/commit/4fdfc4e0cf9994d0d69e2b74daefc36d22a090ce).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54078/
Test FAILed.

vanzin commented:
Ah, mystery solved. While running `dev/mima` those libraries are being copied over and messing the later testing steps. Now need to figure out the proper way of fixing that.

Commit:
Use the same arguments for dev/mima as for the rest of the build.

This makes sure the same profiles are enabled when running sbt again,
so that the dependency set is the same in all invocations of sbt.

SparkQA commented:
**[Test build #54120 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54120/consoleFull)** for PR 11796 at commit [`90660a5`](https://github.com/apache/spark/commit/90660a5dbc7ccdc75c2a31e3ab0e5bdbad56818c).

SparkQA commented:
**[Test build #54120 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54120/consoleFull)** for PR 11796 at commit [`90660a5`](https://github.com/apache/spark/commit/90660a5dbc7ccdc75c2a31e3ab0e5bdbad56818c).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54120/
Test FAILed.

Commit:
Better error message.

SparkQA commented:
**[Test build #54133 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54133/consoleFull)** for PR 11796 at commit [`976ce2c`](https://github.com/apache/spark/commit/976ce2ccd20e91ace2d8efed49cd41292c7cafd5).

SparkQA commented:
**[Test build #54133 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54133/consoleFull)** for PR 11796 at commit [`976ce2c`](https://github.com/apache/spark/commit/976ce2ccd20e91ace2d8efed49cd41292c7cafd5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54133/
Test PASSed.

vanzin commented:
once more with sbt... retest this please

SparkQA commented:
**[Test build #54185 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54185/consoleFull)** for PR 11796 at commit [`976ce2c`](https://github.com/apache/spark/commit/976ce2ccd20e91ace2d8efed49cd41292c7cafd5).

Commit:
Apply some cleanup / fixes from #11701.

Commit:
Avoid copying dependencies again during mima invocation.

SparkQA commented:
**[Test build #54190 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54190/consoleFull)** for PR 11796 at commit [`01d9f0d`](https://github.com/apache/spark/commit/01d9f0db463d94df45b10f5d77fcee8c04660820).

SparkQA commented:
**[Test build #54185 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54185/consoleFull)** for PR 11796 at commit [`976ce2c`](https://github.com/apache/spark/commit/976ce2ccd20e91ace2d8efed49cd41292c7cafd5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54185/
Test PASSed.

SparkQA commented:
**[Test build #54190 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54190/consoleFull)** for PR 11796 at commit [`01d9f0d`](https://github.com/apache/spark/commit/01d9f0db463d94df45b10f5d77fcee8c04660820).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54190/
Test FAILed.

vanzin commented:
retest this please

SparkQA commented:
**[Test build #54227 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54227/consoleFull)** for PR 11796 at commit [`01d9f0d`](https://github.com/apache/spark/commit/01d9f0db463d94df45b10f5d77fcee8c04660820).

SparkQA commented:
**[Test build #54227 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54227/consoleFull)** for PR 11796 at commit [`01d9f0d`](https://github.com/apache/spark/commit/01d9f0db463d94df45b10f5d77fcee8c04660820).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

Commit:
Merge branch 'master' into SPARK-13579

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54227/
Test PASSed.

SparkQA commented:
**[Test build #54243 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54243/consoleFull)** for PR 11796 at commit [`2f09e07`](https://github.com/apache/spark/commit/2f09e07e638e87ed07ede7eedeed7650120af88c).

SparkQA commented:
**[Test build #54243 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54243/consoleFull)** for PR 11796 at commit [`2f09e07`](https://github.com/apache/spark/commit/2f09e07e638e87ed07ede7eedeed7650120af88c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54243/
Test PASSed.

Commit:
Revert "Fix for SPARK-14134 (#11941)."

This reverts commit 3ae4e026f6421fff5a338b622d22c881faf5e447.

vanzin commented:
(just want to test how this behaves without the shading fix)

SparkQA commented:
**[Test build #54328 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54328/consoleFull)** for PR 11796 at commit [`eb828ce`](https://github.com/apache/spark/commit/eb828ce9ec66f6489ff46a7f750dacd6215ab89c).

vanzin commented:
Aside from the above test I'm running, this should be ready for review. @srowen @JoshRosen could you guys take a look?

SparkQA commented:
**[Test build #54328 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54328/consoleFull)** for PR 11796 at commit [`eb828ce`](https://github.com/apache/spark/commit/eb828ce9ec66f6489ff46a7f750dacd6215ab89c).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54328/
Test FAILed.

vanzin commented:
Hmm that's not the test I was expecting to fail. And the error looks very suspicious, as if there's something wrong with that test. Will take a look. retest this pleas

vanzin commented:
retest this please

SparkQA commented:
**[Test build #54336 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54336/consoleFull)** for PR 11796 at commit [`eb828ce`](https://github.com/apache/spark/commit/eb828ce9ec66f6489ff46a7f750dacd6215ab89c).

SparkQA commented:
**[Test build #54336 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54336/consoleFull)** for PR 11796 at commit [`eb828ce`](https://github.com/apache/spark/commit/eb828ce9ec66f6489ff46a7f750dacd6215ab89c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54336/
Test PASSed.

vanzin commented:
Hmm, now tests pass without the fix for  SPARK-14134... retest this please

SparkQA commented:
**[Test build #54358 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54358/consoleFull)** for PR 11796 at commit [`eb828ce`](https://github.com/apache/spark/commit/eb828ce9ec66f6489ff46a7f750dacd6215ab89c).

SparkQA commented:
**[Test build #54358 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54358/consoleFull)** for PR 11796 at commit [`eb828ce`](https://github.com/apache/spark/commit/eb828ce9ec66f6489ff46a7f750dacd6215ab89c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54358/
Test PASSed.

JoshRosen commented:
Nice! I'll try to review this today / tomorrow to get this in soon.

Commit:
Merge branch 'master' into SPARK-13579

SparkQA commented:
**[Test build #54551 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54551/consoleFull)** for PR 11796 at commit [`ef19b3c`](https://github.com/apache/spark/commit/ef19b3c205e16b33ce230ab4cf77fc5cea156e16).

SparkQA commented:
**[Test build #54551 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54551/consoleFull)** for PR 11796 at commit [`ef19b3c`](https://github.com/apache/spark/commit/ef19b3c205e16b33ce230ab4cf77fc5cea156e16).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54551/
Test PASSed.

JoshRosen commented on assembly/pom.xml:
```diff
@@ -69,6 +68,17 @@
+      that the libraries Spark depend on have it available. We'll package the version that Spark
+      uses (14.0.1) which is not the same as Hadoop dependencies, but works.
+    -->
+    <dependency>
```
When and if this turns out to be a problem for user applications, we can consider a different dependency shading strategy in a followup patch. Therefore, I'm fine with this change for now.

RussellSpitzer replied:
This is a problem for the Spark Cassandra Connector. The Cassandra Java Driver requires a 16.0 or greater version of guava. This necessarily means we need to shade now. This was on our roadmap anyway just wanted you to be aware.

JoshRosen commented on bin/spark-class:
```diff
@@ -36,21 +36,20 @@ else
 fi
 
-if [ ! -d "$SPARK_JARS_DIR" ]; then
+if [ ! -d "$SPARK_JARS_DIR" ] && [ -z "$SPARK_TESTING$SPARK_SQL_TESTING" ]; then
```
Is this an idiomatic way to do an `or` check? Looked like a typo to me at first glance, but I now see what it's doing.

vanzin replied:
I can add an explicit `&& [ -z ... ]` if you prefer, just chose the shorter version in this case.

JoshRosen commented on core/src/test/scala/org/apache/spark/util/FileAppenderSuite.scala:
```diff
@@ -201,24 +201,29 @@ class FileAppenderSuite extends SparkFunSuite with BeforeAndAfter with Logging {
 
     // Make sure only logging errors
```
What's the deal with these logging changes? Got a short summary?

vanzin replied:
For some reason these tests failed in one of my runs. And when they failed, they used leave the error level at "ERROR", which basically means all tests that ran after these didn't write any logs.

JoshRosen commented on dev/run-tests.py:
```diff
@@ -371,9 +371,10 @@ def build_apache_spark(build_tool, hadoop_version):
+    build_profiles = get_hadoop_profiles(hadoop_version) + modules.root.build_profile_flags
     set_title_and_block("Detecting binary incompatibilities with MiMa", "BLOCK_MIMA")
-    run_cmd([os.path.join(SPARK_HOME, "dev", "mima")])
+    run_cmd([os.path.join(SPARK_HOME, "dev", "mima")] + build_profiles)
```
Why do we want to pass the Hadoop version / profile to Mima? The old artifact that we're checking against was built against a fixed Hadoop version. Is the goal here to make sure that Spark's Hadoop profile doesn't affect binary compatibility?

vanzin replied:
I added these arguments so that when the mima script would end up re-running the packaging step, it would not copy new files with different versions into the assembly `jars` directory. That was causing test failures earlier if you look at the comment history.

Later I added a flag to disable copying dependencies altogether, and used it in the mima invocation, making this not entirely necessary. But in the end, I think this is still more correct than the previous code - and, as you say, has the nice effect of making sure binary compatibility doesn't change because of different dependency versions.

JoshRosen commented on dev/run-tests.py:
```diff
@@ -571,8 +572,8 @@ def main():
+        detect_binary_inop_with_mima(hadoop_version)
+        # Since we did not build assembly/package before running dev/mima, we need to
         # do it here because the tests still rely on it; see SPARK-13294 for details.
         build_spark_assembly_sbt(hadoop_version)
```
The original motivation for this change was to ensure that MiMa would not accidentally rely on an assembly. Now that we're completely removing assemblies, does it make sense to go back to always unconditionally running `build_spark_assembly_*` before MiMA? Just wondering if this could simplify things a bit and also because this line comment is outdated / confusing now.

vanzin replied:
I'm not sure about how these two really interact; I noticed that `detect_binary_inop_with_mima` would, directly or indirectly, end up running the `package` target for the build. Also note there are still some assemblies that are built, for streaming connectors.

Overall, because of SPARK-9284, we could potentially change all this to run everything in one step (build, tests and potentially even mima checks), although that could potentially make error messages less clear. I'd rather look at that separately though.

JoshRosen commented on examples/pom.xml:
```diff
@@ -325,38 +297,6 @@
           <skip>true</skip>
         </configuration>
       </plugin>
-      <plugin>
```
I guess all of this test JAR stuff is in the root POM anyways, right?

vanzin replied:
Yes, I changed the root pom to make it easy to disable test jar creation (which is not needed for this module, nor the assembly one).

JoshRosen commented on sql/hive/pom.xml:
```diff
@@ -238,36 +238,12 @@
          </executions>
       </plugin>
 
-      <!-- Deploy datanucleus jars to the spark/lib_managed/jars directory -->
```
Glad to see this special-case being removed!

JoshRosen commented on pom.xml:
```diff
@@ -2161,9 +2182,6 @@
-              <!-- At a minimum we must include this to force effective pom generation -->
-              <include>org.spark-project.spark:unused</include>
-
               <include>org.eclipse.jetty:jetty-io</include>
```
Can we safely remove the entire `maven-shade-plugin` configuration here?

vanzin replied:
No, we're still shading Guava and Jetty. The output of a maven build should not contain any jetty jars (I think I checked that but can double check it).

JoshRosen replied:
Which JAR do the relocated Jetty classes wind up in? Spark core's?

vanzin replied:
Yes. The guava ones end up in network-common IIRC.

JoshRosen replied:
Hmm, I don't seem to be able to find those classes after unzipping the JAR in the /jars folder.

JoshRosen commented on project/SparkBuild.scala:
```diff
@@ -263,8 +264,14 @@ object SparkBuild extends PomBuild {
+   * line, so that dev/mima can run without trying to copy these files again and potentially
+   * causing issues.
+   */
+  if (!"false".equals(System.getProperty("copyDependencies"))) {
```
We currently only run MiMa with SBT so I don't think that the MiMa SBT invocation overwriting some output produced by Maven is a concern at this point. However, I appreciate the flag and propose to leave it in place.

vanzin replied:
See previous discussion for why this was needed.

JoshRosen commented on project/SparkBuild.scala:
```diff
@@ -748,6 +739,13 @@ object Java8TestSettings {
   import BuildCommons._
 
+  private val scalaBinaryVersion =
+    if (System.getProperty("scala-2.10") == "true") {
```
Is it possible to read this from the POM rather than relying on a system property to be set? I'm worried that the approach used here will not work if we have to support three Scala versions (2.10, 2.11, and 2.12).

vanzin replied:
I don't really know. This is the same way the property is checked in `Java8TestSettings`, I just followed the pattern.

JoshRosen replied:
Meh, this is fine for now; I can clean it up later if necessary as part of 2.12.

JoshRosen commented:
Took a quick pass and have a few clarifying questions. This looks like it's in good shape overall, though.

Commit:
Merge branch 'master' into SPARK-13579

SparkQA commented:
**[Test build #54652 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54652/consoleFull)** for PR 11796 at commit [`9275ea6`](https://github.com/apache/spark/commit/9275ea6a717028a3f15590150d132903df5d6f15).

SparkQA commented:
**[Test build #54652 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54652/consoleFull)** for PR 11796 at commit [`9275ea6`](https://github.com/apache/spark/commit/9275ea6a717028a3f15590150d132903df5d6f15).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54652/
Test FAILed.

vanzin commented:
@JoshRosen if you want to take a look at the above failure, it seems related to running the build with jdk8. retest this please

JoshRosen commented:
Hmm, that's unfortunate. Let me flip the switch to move the PR builder back to Java 7...

SparkQA commented:
**[Test build #54656 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54656/consoleFull)** for PR 11796 at commit [`9275ea6`](https://github.com/apache/spark/commit/9275ea6a717028a3f15590150d132903df5d6f15).

JoshRosen commented:
Jenkins, retest this please.

SparkQA commented:
**[Test build #54657 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54657/consoleFull)** for PR 11796 at commit [`9275ea6`](https://github.com/apache/spark/commit/9275ea6a717028a3f15590150d132903df5d6f15).

SparkQA commented:
**[Test build #54656 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54656/consoleFull)** for PR 11796 at commit [`9275ea6`](https://github.com/apache/spark/commit/9275ea6a717028a3f15590150d132903df5d6f15).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54656/
Test PASSed.

SparkQA commented:
**[Test build #54657 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54657/consoleFull)** for PR 11796 at commit [`9275ea6`](https://github.com/apache/spark/commit/9275ea6a717028a3f15590150d132903df5d6f15).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54657/
Test PASSed.

Commit:
Merge branch 'master' into SPARK-13579

vanzin commented:
Hi @JoshRosen , any more comments?

SparkQA commented:
**[Test build #54865 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54865/consoleFull)** for PR 11796 at commit [`b5d4292`](https://github.com/apache/spark/commit/b5d4292825834147c874cfc6807af466f7aee732).

SparkQA commented:
**[Test build #54865 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54865/consoleFull)** for PR 11796 at commit [`b5d4292`](https://github.com/apache/spark/commit/b5d4292825834147c874cfc6807af466f7aee732).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54865/
Test PASSed.

JoshRosen commented on bin/spark-class:
```diff
@@ -36,21 +36,20 @@ else
-  SPARK_JARS_DIR="${SPARK_HOME}/lib"
+  SPARK_JARS_DIR="${SPARK_HOME}/jars"
 else
   SPARK_JARS_DIR="${SPARK_HOME}/assembly/target/scala-$SPARK_SCALA_VERSION"
```
I think you also have to append `/jars` here?

Commit:
Add missing "jars" in path.

SparkQA commented:
**[Test build #54892 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54892/consoleFull)** for PR 11796 at commit [`a419707`](https://github.com/apache/spark/commit/a419707a690b30a919dcbf923cd72597e38cfe2f).

JoshRosen commented:
LGTM pending Jenkins. I'll merge this after tests pass and after I wrap up a final local test with the `make-distribution.sh` script.

I'd like to get these core changes in now in order to unblock other build patches and so that they get more widespread community testing. If necessary, we can address small issues in followup patches.

SparkQA commented:
**[Test build #54892 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54892/consoleFull)** for PR 11796 at commit [`a419707`](https://github.com/apache/spark/commit/a419707a690b30a919dcbf923cd72597e38cfe2f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54892/
Test PASSed.

JoshRosen commented:
I also manually ran `DistributedSuite` in SBT to make sure that the SBT build still works; everything looked fine, so I'm merging now. Thanks!

zsxwing commented:
How to build Spark after this patch? Just run `build/sbt clean assembly` but `bin/spark-shell` doesn't work any more:
```
Failed to find Spark jars directory (/Users/zsx/workspace/spark/assembly/target/scala-2.10/jars).
You need to build Spark before running this program.
```

JoshRosen commented:
@zsxwing, try `build/sbt clean package`?

vanzin commented:
Yeah it should be "package" now. I'll send a message to the dev list.

zsxwing commented:
yeah, `package` works. Thanks!
